### DISCUSSION TITLE
实现方式讨论

### DISCUSSION THREAD

**Discussion by mzky** [bddjr/hlfhr#6] [Show and tell]
-

**Comment by bddjr** [comment]
我点进你主页发现你用简体中文，所以我直接在英文后面附加一段中文，这样前面那段老外看，后面那段国人看（doge）

之前的那些方案的bug是，请求稍微长一点就会导致connection was reset，因为请求内容没被服务器读完。这个bug经过这个mod不断的测试迭代，现在已经没有了。

这个mod其它文件不是重点，重点实现在 [conn.go](https://github.com/bddjr/hlfhr/blob/main/conn.go) 和 [conn-http-header-reader.go](https://github.com/bddjr/hlfhr/blob/main/conn-http-header-reader.go) 里面，你可以看看。

**Reply by mzky** [reply]
嗯嗯 刚刚正在看，我之前也是建了一个新库，因为golang版本更新太快 ，我又不怕错过新版特性，后来改成追加函数了

**Reply by bddjr** [reply]
现在这个mod即使go标准库更新了也不会出问题

**Comment by bddjr** [comment]
我现在把这个issue转到讨论区了

**Comment by bddjr** [comment]
我最初也是写个脚本实现，后来发现有bug又改了几次（然而还是有bug）。
https://github.com/bddjr/go-https-port-auto-redirect-for-http/tree/main

后来我觉得那样不是很好，实现一个功能还得以管理员权限改标准库，导致不能直接使用官方的标准库过编译，这太麻烦了。  
于是我就开始研究标准库看看有没有突破口，研究到 tls.NewListener 的时候发现原来可以劫持，于是就有了这个mod。  

最初这个mod还是尝试一次读完所有字节，然后进行简单处理，但是这样显然还是有bug，如果我把请求切割成两个包发送，这个处理器就无法正确处理请求，还是会connection was reset。  

后来我开始研究 bufio.NewBuffer ，发现原来能通过这种方法实现内容复用，于是0.2.x版本就有了。（这个版本号不是很严谨）  

近期，我尝试让http异步处理，这样能在检测到时http请求时让上级立即打印日志，释放上级的部分资源缓存。  
然后我还自己实现了单字节buffer，这样在https握手的时候就不会创建572字节的 bufio.NewBuffer 了，理论上对https的tcp握手请求的性能影响会更小。  
于是有了1.0.0版本

**Reply by mzky** [reply]
你真是太棒了，我都没有关注到 tls.NewListener ，我也抽空看看这个，期待你的新版本
![2461ABDD](https://github.com/user-attachments/assets/059d4d8c-64bf-4a8a-a46d-197047893e67)

**Reply by bddjr** [reply]
以后如果没有bug可能不会有新版本了，现在可能已经是极限了，我想不出有什么优化空间。 ![](https://i0.hdslb.com/bfs/emote/685612eadc33f6bc233776c6241813385844f182.png@32w_32h.avif)

**Reply by bddjr** [reply]
目前性能方面也不是一定没有优化空间，但是没必要继续优化了，这点性能占用还不如http标准库高，这个mod本来就是配合http标准库使用的。

如果不使用Response响应而是直接拼接字符串，性能方面可能是会好一点，但是代码会很杂乱，以前的版本用过拼接字符串的方案，现在已经弃用了。

**Comment by bddjr** [comment]
我又想到了一点点优化空间，下一个小版本1.0.1会改

**Reply by bddjr** [reply]
v1.0.1 已发布 https://github.com/bddjr/hlfhr/releases/tag/v1.0.1

@mzky

**Reply by bddjr** [reply]
新版本进一步减少了对tls握手的性能影响，不会在tls握手时创建chhr，只会在处理http请求时创建chhr。

**Comment by bddjr** [comment]
你有哪里看不懂可以直接问我

**Comment by mzky** [comment]
你需要请求头超过576字节的内容吗？ 我这没有这种需求

**Reply by bddjr** [reply]
tls给你576字节的byte数组不等于它会读满，发现不是tls请求的时候，如果因为网络原因刚好被切片了呢？还有我随便往网址的参数里加点东西，或者域名长一些，很容易就超过这么多字节了。  
此外，现在的浏览器的UserAgent也是越堆越多，无法预测以后的UserAgent会有多长，保证请求头全部被读完是服务端该干的事情。  

go标准库在这块的处理是相当敷衍，因为他们不需要把除了443端口以外的https端口网址发给用户。

**Reply by bddjr** [reply]
说个更现实点的场景，假如我在http设置过cookie，后来服务端升级https了，此时客户端没清掉http保存的cookie，只能一脸懵的看着connection was reset。  
或者我哪天真要在网址里堆参数，直到后来我才在同一个端口升级https，访问原来的网址直接connection was reset了，这好吗这不好。  
作为服务器应该保证的是普通的请求能像原来一样被读取，然后正确地返回重定向响应。

**Reply by bddjr** [reply]
我写这个mod的目标是，让配置端口自动https变得像nginx一样简单，甚至让自定义处理器变得更简单。  

我不知道你有没有看到我举的nginx的例子
```nginx
error_page 497 https://$host:$server_port$request_uri ;
```

**Reply by bddjr** [reply]
我直接说了，我之前在本地用localhost测试遇到过无数次connection was reset了，所以我决定解决这个问题。

**Reply by bddjr** [reply]
（我为什么要和你这杠精解释？）  
## 你说得对，但是新版浏览器默认在前3秒自动访问https协议，你可以不用解决了。

**Reply by mzky** [reply]
我这试了几个浏览器没发现问题，可能是我这只有登录页需要跳转，不需要传的请求头；
标准库嵌套的太多了，并且变量都是私有的，我正在试你的库,这么写是没问题的 
```
package main

import (
	"fmt"
	"github.com/bddjr/hlfhr"
	"github.com/gin-gonic/gin"
	"io"
	"net/http"
	"time"
)

func main() {
	engine := gin.New()
	engine.GET("/login.html", func(c *gin.Context) {
		c.String(http.StatusOK, "Hello World!")
	})
	srv := hlfhr.New(&http.Server{
		Addr:              ":5678",
		Handler:           engine,
		ReadHeaderTimeout: 10 * time.Second,
		IdleTimeout:       10 * time.Second,
	})

	srv.HttpOnHttpsPortErrorHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/html")
		w.WriteHeader(307)
		r.URL.Scheme = "https"
		r.URL.Path = "/login.html"
		r.URL.Host = r.Host
		io.WriteString(w,
			fmt.Sprintf("<script>location.href='%s'</script>", r.URL.String()))
	})

	err := srv.ListenAndServeTLS("cert.pem", "key.pem")
	fmt.Println(err)
}
```
![image](https://github.com/user-attachments/assets/6267246c-5578-48bb-a1ab-f204b1d52bc5)


```
`HTTP/1.1 307 Internal Redirect
Location: url地址
Timing-Allow-Origin: *
Cross-Origin-Resource-Policy: Cross-Origin
Non-Authoritative-Reason: HSTS // 能否提供一个自定义这个HSTS 也是用于跳转的，
x-xss-protection: 0

<head><meta http-equiv="refresh" content="1;url=url地址"></head>`
```
```
io.WriteString(w,这里写html) 无法打开页面
```
看着像是连接close导致的
![image](https://github.com/user-attachments/assets/89999685-8ef2-4736-a6ed-47106b72f33a)

**Reply by mzky** [reply]
> 说个更现实点的场景，假如我在http设置过cookie，后来服务端升级https了，此时客户端没清掉http保存的cookie，只能一脸懵的看着connection was reset。 或者我哪天真要在网址里堆参数，直到后来我才在同一个端口升级https，访问原来的网址直接connection was reset了，这好吗这不好。 作为服务器应该保证的是普通的请求能像原来一样被读取，然后正确地返回重定向响应。

你说的这个cookie倒是有可能

**Reply by bddjr** [reply]
你这需求根本没必要自定义Handler啊，只是浏览器访问的话直接302就行了  
（终于回消息了啊）

**Reply by bddjr** [reply]
你这缺少Location头了😂，建议参考MDN文档

https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Redirections

**Reply by bddjr** [reply]
这和Connection响应头没有任何关系，我还是建议你认真看看MDN文档，直接用bing.com搜对应状态码数字就有了

**Reply by mzky** [reply]
> （我为什么要和你这杠精解释？）
> 
> ## 你说得对，但是新版浏览器默认在前3秒自动访问https协议，你可以不用解决了。

别激动 没有要跟你杠的意思~
![26068A7B](https://github.com/user-attachments/assets/f234e121-d40c-4d6d-a613-d09a4a52070c)

**Reply by bddjr** [reply]
一般301 302 304 307 308状态码的情况下，浏览器可能不认html内容的，只认Location。  

如果要脚本重定向，要用400状态码，或者别的能显示网页内容的状态码

我那个极简响应在你回复我之后的瞬间更改了，我把最新内容复制过来给您看看  

```
"HTTP/1.1 400 Bad Request\r\nConnection: close\r\n\r\n<script>location.protocol='https:'</script>"
```

**Reply by bddjr** [reply]
你应该也发现了，我这个mod会像标准库一样默认带有Date响应头，这会更方便调试区分不同时间的响应

**Reply by mzky** [reply]
> 你这缺少Location头了😂，建议参考MDN文档
> 
> https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Redirections
这个我看过了，所以才用的307，下班了哈，明天再聊，感谢你的项目，用多个浏览器试了没问题
![image](https://github.com/user-attachments/assets/b958f12d-96bf-4a8d-be24-c83f19b143a3)

**Reply by bddjr** [reply]
307状态码可以直接复制README.md里的示范，不需要任何额外修改

https://github.com/bddjr/hlfhr?tab=readme-ov-file#httponhttpsporterrorhandler
```go
// 307 Temporary Redirect
srv.HttpOnHttpsPortErrorHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
	hlfhr.RedirectToHttps(w, r, 307)
})
```

**Reply by bddjr** [reply]
我觉得还是有必要较真一下，你说只需要login.html的重定向，那更不需要自定义handler，默认的302够用了。

302方法与307方法的区别是，302方法可能会导致客户端把 POST 转为 GET ，只有遇到这种问题的时候才可能需要改用307。

**Reply by bddjr** [reply]
我认为你还应该看一下Connection响应头是什么意思  
https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Headers/Connection

**Reply by bddjr** [reply]
使用302状态码的好处是，f12抓包一眼就能看出302是服务器返回的，307是浏览器自己HSTS的。

**Comment by bddjr** [comment]
你可以看看 https://github.com/bddjr/hlfhr/blob/main/response.go#L66 ，这个mod实现302状态码与Location响应头重定向的方法就在这

**Reply by mzky** [reply]
好的 忘了贴这张图 当时选的都是这个页面写的最佳方式。。。
![image](https://github.com/user-attachments/assets/b899e628-55fa-4c44-85b1-3b0c1ef44530)

**Reply by bddjr** [reply]
别 **断章取义** 啊喂！！！    
“HTTP 重定向是创建重定向的最佳方式，**但是**有时候你并**不能控制服务器**。针对这些**特定的应用情景**……”    
这文档说的是像 github.io 这种**静态托管**，不能控制服务器返回302状态码，所以需要用html实现重定向。

**Comment by bddjr** [comment]
你用edit我看不到啊，我没收到邮件啊，如果不是有人叫你重新开启issue我根本不知道你改了消息

还有，请去掉除了Connection Location Content-Length Date以外的响应头，包括用于诊断的那行HSTS，它本该由浏览器生成，而不是由服务器返回。

未登录的路径重定向处理应该交给HTTPS，而不是HTTP处理，因为cookie为了安全起见设置secure标签后，HTTP读不到cookie

**Comment by bddjr** [comment]
把这堆没用的响应头去掉行吗，特别是那行`Non-Authoritative-Reason: HSTS`是非标准响应头，应该在浏览器自己HSTS时才会有，而不是服务器响应
```
Timing-Allow-Origin: *
Cross-Origin-Resource-Policy: Cross-Origin
Non-Authoritative-Reason: HSTS
x-xss-protection: 0
```

**Reply by mzky** [reply]
好

**Comment by bddjr** [comment]
我看了下，你写的那个实现方案是通过劫持handshakeFn实现的，在它返回错误的时候处理Write，这导致我不能完整获取请求内容，意味着不能依据请求内容响应。  

你这ServeTLS后面有些行根本没必要自己弄一遍，交给标准库就可以了，还有你这名称是不是错了，应该是ListenAndServeTLS

还有，别用内网IP测试，用localhost测试出现connection was reset的概率会更高。

**Comment by bddjr** [comment]
说实话啊，你的实现方式还是别往issue里发了，没有任何帮助

**Comment by bddjr** [comment]
我也写了个简单版（不支持自定义handler）  
https://github.com/bddjr/hlfhr/tree/main/simplehlfhr

**Comment by bddjr** [comment]
你的实现方式有个bug：http2没有被正确的初始化，浏览器进行TLS握手成功后会使用http/1.1，这影响了传输效率。

**Reply by bddjr** [reply]
我用相同的方式实现，结果h2就被干没了  
https://github.com/bddjr/hlfhr/tree/tlslistener
